src/proxy/http/profiles.rs

```diff
@@ -0,0 +1,389 @@
+#![allow(dead_code)]
+
+extern crate tower_discover;
+
+use futures::Stream;
+use http;
+use indexmap::IndexMap;
+use regex::Regex;
+use std::iter::FromIterator;
+use std::sync::Arc;
+use std::{error, fmt};
+
+use transport::DnsNameAndPort;
+
+pub trait CanGetDestination {
+    fn get_destination(&self) -> Option<&DnsNameAndPort>;
+}
+
+pub type Routes = Vec<(RequestMatch, Route)>;
+
+pub trait GetRoutes {
+    type Stream: Stream<Item = Routes, Error = Error>;
+
+    fn get_routes(&self, dst: &DnsNameAndPort) -> Option<Self::Stream>;
+}
+
+pub trait WithRoute {
+    type Output;
+
+    fn with_route(self, route: Route) -> Self::Output;
+}
+
+#[derive(Debug)]
+pub enum Error {}
+
+#[derive(Clone, Debug, Default)]
+pub struct Route {
+    labels: Arc<IndexMap<String, String>>,
+    response_classes: Arc<Vec<ResponseClass>>,
+}
+
+#[derive(Clone, Debug)]
+pub enum RequestMatch {
+    All(Vec<RequestMatch>),
+    Any(Vec<RequestMatch>),
+    Not(Box<RequestMatch>),
+    Path(Regex),
+    Method(http::Method),
+}
+
+#[derive(Clone, Debug)]
+pub struct ResponseClass {
+    is_failure: bool,
+    match_: ResponseMatch,
+}
+
+#[derive(Clone, Debug)]
+pub enum ResponseMatch {
+    All(Vec<ResponseMatch>),
+    Any(Vec<ResponseMatch>),
+    Not(Box<ResponseMatch>),
+    Status {
+        min: http::StatusCode,
+        max: http::StatusCode,
+    },
+}
+
+// === impl Route ===
+
+impl Route {
+    pub fn new<I>(label_iter: I, response_classes: Vec<ResponseClass>) -> Self
+    where
+        I: Iterator<Item = (String, String)>,
+    {
+        let labels = {
+            let mut pairs = label_iter.collect::<Vec<(String, String)>>();
+            pairs.sort_by(|(k0, _), (k1, _)| k0.cmp(k1));
+            Arc::new(IndexMap::from_iter(pairs))
+        };
+
+        Self {
+            labels,
+            response_classes: response_classes.into(),
+        }
+    }
+
+    pub fn labels(&self) -> &Arc<IndexMap<String, String>> {
+        &self.labels
+    }
+
+    pub fn response_classes(&self) -> &Arc<Vec<ResponseClass>> {
+        &self.response_classes
+    }
+}
+
+// === impl RequestMatch ===
+
+impl RequestMatch {
+    fn is_match<B>(&self, req: &http::Request<B>) -> bool {
+        match self {
+            RequestMatch::Method(ref method) => req.method() == *method,
+            RequestMatch::Path(ref re) => re.is_match(req.uri().path()),
+            RequestMatch::Not(ref m) => !m.is_match(req),
+            RequestMatch::All(ref matches) => {
+                for ref m in matches {
```

Comment:
TIOLI: I think this can just be:
```suggestion
                matches.iter().all(RequestMatch::is_match),
```